File: superset-frontend/src/hooks/apiResources/schemas.test.ts
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import rison from 'rison';
import fetchMock from 'fetch-mock';
import { act, renderHook, waitFor } from '@testing-library/react';
import {
  createWrapper,
  defaultStore as store,
} from 'spec/helpers/testing-library';
import { api } from 'src/hooks/apiResources/queryApi';
import { useSchemas } from './schemas';

const fakeApiResult = {
  result: ['test schema 1', 'test schema b'],
};
const fakeApiResult2 = {
  result: ['test schema 2', 'test schema a'],
};
const fakeApiResult3 = {
  result: ['test schema 3', 'test schema c'],
};

const expectedResult = fakeApiResult.result.map((value: string) => ({
  value,
  label: value,
  title: value,
}));
const expectedResult2 = fakeApiResult2.result.map((value: string) => ({
  value,
  label: value,
  title: value,
}));
const expectedResult3 = fakeApiResult3.result.map((value: string) => ({
  value,
  label: value,
  title: value,
}));

// eslint-disable-next-line no-restricted-globals -- TODO: Migrate from describe blocks
describe('useSchemas hook', () => {
  beforeEach(() => {
    fetchMock.clearHistory().removeRoutes();
    store.dispatch(api.util.resetApiState());
  });

  test('returns api response mapping json result', async () => {
    const expectDbId = 'db1';
    const forceRefresh = false;
    const schemaApiRoute = `glob:*/api/v1/database/${expectDbId}/schemas/*`;
    fetchMock.get(schemaApiRoute, fakeApiResult);
    const onSuccess = jest.fn();
    const { result } = renderHook(
      () =>
        useSchemas({
          dbId: expectDbId,
          onSuccess,
        }),
      {
        wrapper: createWrapper({
          useRedux: true,
          store,
        }),
      },
    );
    await waitFor(() =>
      expect(fetchMock.callHistory.calls(schemaApiRoute).length).toBe(1),
    );
    expect(result.current.data).toEqual(expectedResult);
    expect(
      fetchMock.callHistory.calls(
        `end:/api/v1/database/${expectDbId}/schemas/?q=${rison.encode({
          force: forceRefresh,
        })}`,
      ).length,
    ).toBe(1);
    expect(onSuccess).toHaveBeenCalledTimes(1);
    act(() => {
      result.current.refetch();
    });
    await waitFor(() =>
      expect(fetchMock.callHistory.calls(schemaApiRoute).length).toBe(2),
    );
    expect(
      fetchMock.callHistory.calls(
        `end:/api/v1/database/${expectDbId}/schemas/?q=${rison.encode({
          force: true,
        })}`,
      ).length,
    ).toBe(1);
    expect(onSuccess).toHaveBeenCalledTimes(2);
    expect(result.current.data).toEqual(expectedResult);
  });

  test('returns cached data without api request', async () => {
    const expectDbId = 'db1';
    const schemaApiRoute = `glob:*/api/v1/database/${expectDbId}/schemas/*`;
    fetchMock.get(schemaApiRoute, fakeApiResult);
    const { result, rerender } = renderHook(
      () =>
        useSchemas({
          dbId: expectDbId,
        }),
      {
        wrapper: createWrapper({
          useRedux: true,
          store,
        }),
      },
    );
    await waitFor(() => expect(result.current.data).toEqual(expectedResult));
    expect(fetchMock.callHistory.calls(schemaApiRoute).length).toBe(1);
    rerender();
    await waitFor(() => expect(result.current.data).toEqual(expectedResult));
    expect(fetchMock.callHistory.calls(schemaApiRoute).length).toBe(1);
  });

  test('returns refreshed data after expires', async () => {
    const expectDbId = 'db1';
    const schemaApiRoute = `glob:*/api/v1/database/*/schemas/*`;
    fetchMock.get(schemaApiRoute, ({ url }) =>
      url.includes(expectDbId) ? fakeApiResult : fakeApiResult2,
    );
    const onSuccess = jest.fn();
    const { result, rerender } = renderHook(
      ({ dbId }) =>
        useSchemas({
          dbId,
          onSuccess,
        }),
      {
        initialProps: { dbId: expectDbId },
        wrapper: createWrapper({
          useRedux: true,
          store,
        }),
      },
    );

    await waitFor(() =>
      expect(result.current.currentData).toEqual(expectedResult),
    );
    expect(fetchMock.callHistory.calls(schemaApiRoute).length).toBe(1);
    expect(onSuccess).toHaveBeenCalledTimes(1);

    rerender({ dbId: 'db2' });
    await waitFor(() =>
      expect(result.current.currentData).toEqual(expectedResult2),
    );
    expect(fetchMock.callHistory.calls(schemaApiRoute).length).toBe(2);
    expect(onSuccess).toHaveBeenCalledTimes(2);

    rerender({ dbId: expectDbId });
    await waitFor(() =>
      expect(result.current.currentData).toEqual(expectedResult),
    );
    expect(fetchMock.callHistory.calls(schemaApiRoute).length).toBe(2);
    expect(onSuccess).toHaveBeenCalledTimes(2);

    // clean up cache
    act(() => {
      store.dispatch(api.util.invalidateTags(['Schemas']));
    });

    // Only the currently subscribed query (expectDbId) is refetched on
    // invalidation; the previously visited db2 entry is no longer subscribed.
    await waitFor(() =>
      expect(fetchMock.callHistory.calls(schemaApiRoute).length).toBe(3),
    );
    expect(fetchMock.callHistory.calls(schemaApiRoute)[2].url).toContain(
      expectDbId,
    );
    await waitFor(() =>
      expect(result.current.currentData).toEqual(expectedResult),
    );
  });

  test('fires onSuccess when the subscribed query refetches after invalidateTags', async () => {
    // Regression test for the OAuth2 retry-after-redirect path (PR #41101).
    // The redirect handler dispatches invalidateTags, which refetches the
    // SUBSCRIBED query (not the lazy trigger). onSuccess must still fire so
    // consumers holding local state (e.g. an auth error banner) get cleared.
    const expectDbId = 'db1';
    const schemaApiRoute = `glob:*/api/v1/database/${expectDbId}/schemas/*`;
    fetchMock.get(schemaApiRoute, fakeApiResult);
    const onSuccess = jest.fn();
    const { result } = renderHook(
      () =>
        useSchemas({
          dbId: expectDbId,
          onSuccess,
        }),
      {
        wrapper: createWrapper({
          useRedux: true,
          store,
        }),
      },
    );

    await waitFor(() =>
      expect(fetchMock.callHistory.calls(schemaApiRoute).length).toBe(1),
    );
    expect(result.current.currentData).toEqual(expectedResult);
    expect(onSuccess).toHaveBeenCalledTimes(1);

    act(() => {
      store.dispatch(
        api.util.invalidateTags([{ type: 'Schemas', id: 'LIST' }]),
      );
    });

    await waitFor(() =>
      expect(fetchMock.callHistory.calls(schemaApiRoute).length).toBe(2),
    );
    await waitFor(() => expect(onSuccess).toHaveBeenCalledTimes(2));
    // isRefetched must be false so the selectors don't emit a "List refreshed"
    // toast for an automatic refetch the user did not request.
    expect(onSuccess).toHaveBeenLastCalledWith(expectedResult, false);
  });

  test('recovers from an error when the subscribed query refetches (OAuth2 retry)', async () => {
    const expectDbId = 'db1';
    const schemaApiRoute = `glob:*/api/v1/database/${expectDbId}/schemas/*`;
    let shouldFail = true;
    fetchMock.get(schemaApiRoute, () =>
      shouldFail ? { status: 500, body: {} } : fakeApiResult,
    );
    const onSuccess = jest.fn();
    const onError = jest.fn();
    renderHook(
      () =>
        useSchemas({
          dbId: expectDbId,
          onSuccess,
          onError,
        }),
      {
        wrapper: createWrapper({
          useRedux: true,
          store,
        }),
      },
    );

    await waitFor(() => expect(onError).toHaveBeenCalledTimes(1));
    expect(onSuccess).not.toHaveBeenCalled();

    // The OAuth2 redirect completes and the token is stored: the next fetch
    // succeeds, and onSuccess must fire to clear the error banner.
    shouldFail = false;
    act(() => {
      store.dispatch(
        api.util.invalidateTags([{ type: 'Schemas', id: 'LIST' }]),
      );
    });

    await waitFor(() => expect(onSuccess).toHaveBeenCalledTimes(1));
    expect(onSuccess).toHaveBeenLastCalledWith(expectedResult, false);
  });

  test('returns correct schema list by a catalog', async () => {
    const dbId = '1';
    const expectCatalog = 'catalog3';
    const schemaApiRoute = `glob:*/api/v1/database/*/schemas/*`;
    fetchMock.get(schemaApiRoute, ({ url }) =>
      url.includes(`catalog:${expectCatalog}`)
        ? fakeApiResult3
        : fakeApiResult2,
    );
    const onSuccess = jest.fn();
    const { result, rerender } = renderHook(
      ({ dbId, catalog }) =>
        useSchemas({
          dbId,
          catalog,
          onSuccess,
        }),
      {
        initialProps: { dbId, catalog: expectCatalog },
        wrapper: createWrapper({
          useRedux: true,
          store,
        }),
      },
    );

    await waitFor(() =>
      expect(fetchMock.callHistory.calls(schemaApiRoute).length).toBe(1),
    );
    expect(result.current.data).toEqual(expectedResult3);
    expect(onSuccess).toHaveBeenCalledTimes(1);

    rerender({ dbId, catalog: 'catalog2' });
    await waitFor(() =>
      expect(fetchMock.callHistory.calls(schemaApiRoute).length).toBe(2),
    );
    expect(result.current.data).toEqual(expectedResult2);

    rerender({ dbId, catalog: expectCatalog });
    expect(result.current.data).toEqual(expectedResult3);
    expect(fetchMock.callHistory.calls(schemaApiRoute).length).toBe(2);
  });
});

File: tests/unit_tests/semantic_layers/schemas_test.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

import pytest
from marshmallow import ValidationError

from superset.semantic_layers.schemas import (
    SemanticLayerPostSchema,
    SemanticLayerPutSchema,
    SemanticViewPutSchema,
)


def test_semantic_view_put_schema_both_fields() -> None:
    """Test loading both description and cache_timeout."""
    schema = SemanticViewPutSchema()
    result = schema.load({"description": "A description", "cache_timeout": 300})
    assert result == {"description": "A description", "cache_timeout": 300}


def test_semantic_view_put_schema_description_only() -> None:
    """Test loading with only description."""
    schema = SemanticViewPutSchema()
    result = schema.load({"description": "Just a description"})
    assert result == {"description": "Just a description"}


def test_semantic_view_put_schema_cache_timeout_only() -> None:
    """Test loading with only cache_timeout."""
    schema = SemanticViewPutSchema()
    result = schema.load({"cache_timeout": 600})
    assert result == {"cache_timeout": 600}


def test_semantic_view_put_schema_empty() -> None:
    """Test loading empty payload."""
    schema = SemanticViewPutSchema()
    result = schema.load({})
    assert result == {}


def test_semantic_view_put_schema_null_description() -> None:
    """Test that description accepts None."""
    schema = SemanticViewPutSchema()
    result = schema.load({"description": None})
    assert result == {"description": None}


def test_semantic_view_put_schema_null_cache_timeout() -> None:
    """Test that cache_timeout accepts None."""
    schema = SemanticViewPutSchema()
    result = schema.load({"cache_timeout": None})
    assert result == {"cache_timeout": None}


def test_semantic_view_put_schema_invalid_cache_timeout() -> None:
    """Test that non-integer cache_timeout raises ValidationError."""
    schema = SemanticViewPutSchema()
    with pytest.raises(ValidationError) as exc_info:
        schema.load({"cache_timeout": "not_a_number"})
    assert "cache_timeout" in exc_info.value.messages


def test_semantic_view_put_schema_unknown_field() -> None:
    """Test that unknown fields raise ValidationError."""
    schema = SemanticViewPutSchema()
    with pytest.raises(ValidationError) as exc_info:
        schema.load({"unknown_field": "value"})
    assert "unknown_field" in exc_info.value.messages


# =============================================================================
# SemanticLayerPostSchema tests
# =============================================================================


def test_post_schema_all_fields() -> None:
    """Test loading all fields."""
    schema = SemanticLayerPostSchema()
    result = schema.load({
        "name": "My Layer",
        "description": "A layer",
        "type": "snowflake",
        "configuration": {"account": "test"},
        "cache_timeout": 300,
    })
    assert result["name"] == "My Layer"
    assert result["type"] == "snowflake"
    assert result["configuration"] == {"account": "test"}
    assert result["cache_timeout"] == 300


def test_post_schema_required_fields_only() -> None:
    """Test loading with only required fields."""
    schema = SemanticLayerPostSchema()
    result = schema.load({
        "name": "My Layer",
        "type": "snowflake",
        "configuration": {"account": "test"},
    })
    assert result["name"] == "My Layer"
    assert "description" not in result
    assert "cache_timeout" not in result


def test_post_schema_missing_name() -> None:
    """Test that missing name raises ValidationError."""
    schema = SemanticLayerPostSchema()
    with pytest.raises(ValidationError) as exc_info:
        schema.load({"type": "snowflake", "configuration": {}})
    assert "name" in exc_info.value.messages


def test_post_schema_missing_type() -> None:
    """Test that missing type raises ValidationError."""
    schema = SemanticLayerPostSchema()
    with pytest.raises(ValidationError) as exc_info:
        schema.load({"name": "My Layer", "configuration": {}})
    assert "type" in exc_info.value.messages


def test_post_schema_missing_configuration() -> None:
    """Test that missing configuration raises ValidationError."""
    schema = SemanticLayerPostSchema()
    with pytest.raises(ValidationError) as exc_info:
        schema.load({"name": "My Layer", "type": "snowflake"})
    assert "configuration" in exc_info.value.messages


def test_post_schema_null_description() -> None:
    """Test that description accepts None."""
    schema = SemanticLayerPostSchema()
    result = schema.load({
        "name": "My Layer",
        "type": "snowflake",
        "configuration": {},
        "description": None,
    })
    assert result["description"] is None


# =============================================================================
# SemanticLayerPutSchema tests
# =============================================================================


def test_put_schema_all_fields() -> None:
    """Test loading all fields."""
    schema = SemanticLayerPutSchema()
    result = schema.load({
        "name": "Updated",
        "description": "New desc",
        "configuration": {"account": "new"},
        "cache_timeout": 600,
    })
    assert result["name"] == "Updated"
    assert result["configuration"] == {"account": "new"}


def test_put_schema_empty() -> None:
    """Test loading empty payload."""
    schema = SemanticLayerPutSchema()
    result = schema.load({})
    assert result == {}


def test_put_schema_name_only() -> None:
    """Test loading with only name."""
    schema = SemanticLayerPutSchema()
    result = schema.load({"name": "New Name"})
    assert result == {"name": "New Name"}


def test_put_schema_configuration_only() -> None:
    """Test loading with only configuration."""
    schema = SemanticLayerPutSchema()
    result = schema.load({"configuration": {"key": "value"}})
    assert result == {"configuration": {"key": "value"}}


def test_put_schema_unknown_field() -> None:
    """Test that unknown fields raise ValidationError."""
    schema = SemanticLayerPutSchema()
    with pytest.raises(ValidationError) as exc_info:
        schema.load({"unknown_field": "value"})
    assert "unknown_field" in exc_info.value.messages
